solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java

```diff
@@ -488,4 +468,35 @@ public interface CollectionStateProps {
     String SHARDS = "shards";
     String PER_REPLICA_STATE = "perReplicaState";
   }
+
+  public static class PrsSupplier implements Supplier<PerReplicaStates> {
+
+    protected volatile PerReplicaStates prs;
+
+    PrsSupplier() {}
+
+    PrsSupplier(PerReplicaStates prs) {
+      this.prs = prs;
+    }
+
+    @Override
+    public PerReplicaStates get() {
+      return prs;
+    }
+  }
+
+  public static final ThreadLocal<DocCollection.PrsSupplier> REPLICASTATES_PROVIDER =
```

Comment:
Quick question: is there any alternative to avoid threadlocal. Give me some more context on this

Reply:
Yes, there is. The  only reason why I used the ThreadLocal Option was to minimize the changes. However, I think I should explore that route